R/Data-methods.R

```diff
@@ -415,6 +417,84 @@ setMethod(
   }
 )
 
+# DataDA-update ----
+
+#' Updating `DataDA` Objects
+#'
+#' @description `r lifecycle::badge("stable")`
+#'
+#' A method that updates existing [`DataDA`] object with new data.
+#'
+#' @note This function is capable of not only adding new patients but also
+#'   updates existing ones with respect to `y`, `t0`, `u` slots.
+#'
+#' @param object (`DataDA`)\cr object you want to update.
+#' @param u (`numeric`)\cr the new DLT free survival times for all patients,
+#'   i.e. for existing patients in the `object` as well as for new patients.
+#' @param t0 (`numeric`)\cr the time that each patient starts DLT observation window.
+#'   This parameter covers all patients, i.e. existing patients in the `object`
+#'   as well as for new patients.
+#' @param trialtime (`numeric`)\cr current time in the trial, i.e. a followup time.
+#' @param y (`numeric`)\cr the new DLTs for all patients, i.e. for existing
+#'   patients in the `object` as well as for new patients.
+#' @param ... further arguments passed to `Data` update method [`update-Data-method`].
+#'   These are is used when there are new patients to be added to the cohort.
+#' @param check (`flag`)\cr whether the validation of the updated object
+#'   should be conducted. See help for [`update-Data-method`] for more details
+#'   on the use case of this parameter.
+#'
+#' @return The new, updated [`DataDA`] object.
+#'
+#' @aliases update-DataDA-method
+#' @example examples/Data-method-update-DataDA.R
+#' @export
+#'
+setMethod(
+  f = "update",
+  signature = signature(object = "DataDA"),
+  definition = function(object,
+                        u,
+                        t0,
+                        trialtime,
+                        y,
+                        ...,
+                        check = TRUE) {
+    assert_numeric(y, lower = 0, upper = 1)
+    assert_true(length(y) == 0 || length(y) >= object@nObs)
+    assert_numeric(u, lower = 0, len = length(y))
+    assert_numeric(t0, lower = 0, len = length(y))
+    assert_numeric(trialtime, lower = max(object@t0))
+    assert_true(length(trialtime) == 0L || length(trialtime) == 1L)
+    assert_flag(check)
+
+    # How many additional patients.
+    n <- max(length(y) - object@nObs, 0L)
+
+    # Update slots corresponding to `Data` class.
+    object <- callNextMethod(
+      object = object,
+      y = y[object@nObs + seq_len(n)], # Empty vector when n = 0.
+      ...,
+      check = FALSE
+    )
+
+    # DLT will be observed once the followup time >= the time to DLT.
+    object@y <- safeInteger(y * (trialtime >= t0 + u))
```

Comment:
please adjust comment - we only observe DLT when both time condition is fulfilled and as well in addition `y = 1`